### PR TITLE
feat: improve arch-specific code similarity

### PR DESCRIPTION
**Pull request by sunnygleason**
* This replaces #46
* If this looks good, we can merge to 0.2 and I'll put together another PR to track a reasonable upstream release (which will take care of #62)!

### CODE REVIEW

File: src/avx2/stage1.rs [hunk 1]
```diff
@@ -47,22 +43,21 @@ unsafe fn check_utf8(
...
+        *previous = check_utf8_bytes(input.v0, &previous, has_error);
+        *previous = check_utf8_bytes(input.v1, &previous, has_error);
     }
 }
 
 /// a straightforward comparison of a mask against input. 5 uops; would be
 /// cheaper in AVX512.
 #[cfg_attr(not(feature = "no-inline"), inline(always))]
 fn cmp_mask_against_input(input: &SimdInput, m: u8) -> u64 {
+    #[allow(clippy::cast_possible_wrap, clippy::cast_sign_loss)]
```
Comment (Licenser) [inline src/avx2/stage1.rs]: Something I found out recently: moving this infront of the function would prevent tarpaulin from falsly marking this as a non covered line.

File: src/avx2/stage1.rs [hunk 2]
```diff
@@ -71,12 +66,11 @@ fn cmp_mask_against_input(input: &SimdInput, m: u8) -> u64 {
 // find all values less than or equal than the content of maxval (using unsigned arithmetic)
 #[cfg_attr(not(feature = "no-inline"), inline(always))]
 fn unsigned_lteq_against_input(input: &SimdInput, maxval: __m256i) -> u64 {
+    #[allow(clippy::cast_sign_loss)]
```
Comment (Licenser) [inline src/avx2/stage1.rs]: see above

File: src/avx2/utf8check.rs [hunk 3]
```diff
@@ -68,41 +68,37 @@ fn avxcontinuation_lengths(high_nibbles: __m256i) -> __m256i {
...
-            _mm256_set1_epi8(1),
-        );
-        let sum: __m256i = _mm256_add_epi8(initial_lengths, right1);
-        let right2: __m256i = _mm256_subs_epu8(
-            push_last_2bytes_of_a_to_b(previous_carries, sum),
-            _mm256_set1_epi8(2),
-        );
-        _mm256_add_epi8(sum, right2)
-    }
+unsafe fn carry_continuations(initial_lengths: __m256i, previous_carries: __m256i) -> __m256i {
```
Comment (Licenser) [inline src/avx2/utf8check.rs]: nice spot on pulling the unsafe out and into the function! :D

File: src/neon/stage1.rs [hunk 4]
```diff
@@ -450,10 +447,10 @@ pub fn find_bs_bits_and_quote_bits(v0: uint8x16_t, v1: uint8x16_t) -> ParseStrin
         let sum0 = vpaddq_u8(sum0, sum1);
         let sum0 = vpaddq_u8(sum0, sum0);
 
-        ParseStringHelper {
-            bs_bits: vgetq_lane_u32(vreinterpretq_u32_u8(sum0), 0),
-            quote_bits: vgetq_lane_u32(vreinterpretq_u32_u8(sum0), 1),
-        }
+        (
```
Comment (Licenser) [inline src/neon/stage1.rs]: nice simplification!

File: src/neon/utf8check.rs [hunk 5]
```diff
@@ -178,7 +178,7 @@ fn check_overlong(
     }
 }
 
-pub struct ProcessedUtfBytes {
+pub(crate) struct ProcessedUtfBytes {
```
Comment (Licenser) [inline src/neon/utf8check.rs]: excellent call on locking this down to crate only :)

Comment (sunnygleason) [reply src/neon/utf8check.rs]: just copying the other module which was different! 😂

File: src/avx2/stage1.rs (hunk 1 above)
Comment (sunnygleason) [reply src/avx2/stage1.rs]: nifty! so is this good or should I hoist it up to line 54?

Comment (Licenser) [reply src/avx2/stage1.rs]: It's fine there, we got to go over stuff eventually anyway when we take another crack at coverage :) I just wanted to share it

**Comment by sunnygleason** [comment]
Very cool! I'll aim to do a pass to extract duplicate arch-independent functions this weekend. Thank you again!

**Pull request merged**